jupyterhub_traefik_proxy/toml.py

```diff
@@ -0,0 +1,300 @@
+"""Traefik implementation
+
+Custom proxy implementations can subclass :class:`Proxy`
+and register in JupyterHub config:
+
+.. sourcecode:: python
+
+    from mymodule import MyProxy
+    c.JupyterHub.proxy_class = MyProxy
+
+Route Specification:
+
+- A routespec is a URL prefix ([host]/path/), e.g.
+  'host.tld/path/' for host-based routing or '/path/' for default routing.
+- Route paths should be normalized to always start and end with '/'
+"""
+
+# Copyright (c) Jupyter Development Team.
+# Distributed under the terms of the Modified BSD License.
+
+import json
+import os
+import asyncio
+
+from urllib.parse import urlparse
+from traitlets import Any, default, Unicode
+
+from . import traefik_utils
+from jupyterhub.proxy import Proxy
+from jupyterhub_traefik_proxy import TraefikProxy
+
+
+class TraefikTomlProxy(TraefikProxy):
+    """JupyterHub Proxy implementation using traefik and toml config file"""
+
+    mutex = Any()
+
+    @default("mutex")
+    def _default_mutex(self):
+        return asyncio.Lock()
```

Comment:
The atomic writing should probably follow the pattern [used in the notebook](https://github.com/jupyter/notebook/blob/5.7.0/notebook/services/contents/fileio.py#L71) where we write a temporary file then move it into place. This way there's never a partial configuration file on disk, which traefik might read while we aren't done writing.

It can be a little simpler than the one in the notebook since we know:

1. that it's not Windows
2. that it's a text file
3. we require Python 3
4. we don't need to turn errors into log messages (raising is ok)

Reply:
That doesn't mean we don't want this mutex, I realize. We still want that to protect access to the routes with this.